community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputNodeCacher.java

```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.unsafe.impl.batchimport.input;
+
+import java.io.IOException;
+
+import org.neo4j.io.fs.StoreChannel;
+
+import static org.neo4j.helpers.ArrayUtil.contains;
+
+/**
+ * Caches {@link InputNode} to disk using a binary format.
+ */
+public class InputNodeCacher extends InputEntityCacher<InputNode>
+{
+    private String[] previousLabels = InputEntity.NO_LABELS;
+
+    public InputNodeCacher( StoreChannel channel, StoreChannel header, int bufferSize ) throws IOException
+    {
+        super( channel, header, bufferSize, 1 );
+    }
+
+    @Override
+    protected void writeEntity( InputNode node ) throws IOException
+    {
+        // properties
+        super.writeEntity( node );
+
+        // group
+        writeGroup( node.group(), 0 );
+
+        // id
+        writeValue( node.id() );
+
+        // labels (diff from previous node)
+        if ( node.hasLabelField() )
+        {
+            channel.put( (byte) 3 );
+            channel.putLong( node.labelField() );
+        }
+        else
+        {
+            String[] labels = node.labels();
+            writeDiff( 1, previousLabels, labels ); // removal
```

Comment:
Maybe also create constants for this marker values and use them in both cacher and reader?